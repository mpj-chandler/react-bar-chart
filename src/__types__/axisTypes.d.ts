import { SeriesData, NumericDataPoint } from './seriesTypes';
import { Padding } from './styling';
import { Placement } from '../enums/Placement';

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
    titlePos?: {
        xPos: number,
        yPos: number
    }
}

export interface AxisTickProps {
    data: SeriesData[];
    padding: Padding;
    config?: AxisConfig;
}

