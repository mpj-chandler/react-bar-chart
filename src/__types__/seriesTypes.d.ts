import DataType from '../enums/DataType';
import { Padding } from './styling';
import { AxisRange } from './axisTypes';
import { Placement } from '../enums/Placement';

export interface DataPoint {
    tag?: string;
    y0?: number | null;
    y1: number | null;
}

export interface NamedDataPoint extends DataPoint {
    x: string;    
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
}

export interface NumericDataPoint extends DataPoint {
    x: number;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}