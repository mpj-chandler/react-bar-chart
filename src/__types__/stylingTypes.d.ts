interface Padding {
    left: number;
    right: number;
    top: number;
    bottom: number;
}

export default Padding;

