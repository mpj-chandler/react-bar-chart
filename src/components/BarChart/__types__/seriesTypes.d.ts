import DataType from "../enums/DataType";
import { Padding } from "./stylingTypes";
import { AxisRange } from "./axisTypes";
import Placement from "../enums/Placement";

export interface DataPoint {
    tag?: string;
}

export interface NamedDataPoint extends DataPoint {
    x: string;
    y: number;
}

export interface DateIndexedDataPoint extends DataPoint {
    x: Date;
    y: number;
}

export interface NumericDataPoint extends DataPoint {
    x: number | null;
    y: number | null;
}

export interface NonNullNumericDataPoint extends NumericDataPoint {
    x: number;
    y: number;
}

export interface SeriesData {
    seriesName: string;
    type: {
        x: DataType,
        y: DataType,
    };
    points: Array<NamedDataPoint | DateIndexedDataPoint | NumericDataPoint | NonNullNumericDataPoint>;
}

export interface SeriesConfig {
    numSeries: number,
    seriesIndex: number,
    padding: Padding,
    dataType: DataType,
    xRange: AxisRange,
    yRange: any,
    placement: Placement
}