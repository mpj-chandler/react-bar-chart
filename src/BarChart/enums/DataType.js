"use strict";
exports.__esModule = true;
var DataType;
(function (DataType) {
    DataType["Named"] = "NAMED";
    DataType["DateIndexed"] = "DATE_INDEXED";
    DataType["Numeric"] = "NUMERIC";
    DataType["NonNullNumeric"] = "NON_NULL_NUMERIC";
})(DataType || (DataType = {}));
;
exports["default"] = DataType;
