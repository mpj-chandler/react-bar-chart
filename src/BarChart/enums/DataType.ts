enum DataType {
    Named = 'NAMED',
    DateIndexed = 'DATE_INDEXED',
    Numeric = 'NUMERIC',
    NonNullNumeric = 'NON_NULL_NUMERIC',
};

export default DataType;
