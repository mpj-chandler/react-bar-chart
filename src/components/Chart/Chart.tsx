import React from 'react';
import styles from './Chart.scss';
import XAxis from '../XAxis/XAxis';
import YAxis from '../YAxis/YAxis';

import BarPlot from '../BarPlot/BarPlot';
import { SeriesData, DataPoint } from '../../__types__/seriesTypes';
import { Padding } from '../../__types__/styling';
import { AxisConfig } from '../../__types__/axisTypes';
import AnimationEasingType from '../../enums/AnimationEasingFunction';
import ChartType from '../../enums/ChartType';
import stackSeries from '../StackedBarPlot/StackedBarPlot.mapper';
import StackedBarPlot from '../StackedBarPlot/StackedBarPlot';

export interface BarChartProps {
    data: SeriesData[];
    type: ChartType;
    title: string;
    padding: Padding;
    xAxisConfig?: AxisConfig;
    yAxisConfig?: AxisConfig;
    fillFormatter: (seriesIndex: number, dataPoint: DataPoint, index: number) => string;
    animationEasingType: AnimationEasingType;
}

const Chart: React.FC<BarChartProps> = (props: BarChartProps) => {
    const renderPlot = () => {
        if (props.type === ChartType.BarChart) {
            return (
                <BarPlot
                    data={props.data}
                    padding={props.padding}
                    xAxisConfig={props.xAxisConfig}
                    yAxisConfig={props.yAxisConfig}
                    animationEasingFunction={props.animationEasingType}
                    fillFormatter={props.fillFormatter}
                />
            );
        }

        return (
            <StackedBarPlot
                data={stackSeries(props.data)}
                padding={props.padding}
                xAxisConfig={props.xAxisConfig}
                yAxisConfig={props.yAxisConfig}
                animationEasingFunction={props.animationEasingType}
                fillFormatter={props.fillFormatter}
            />
        );
    }

    return (
        <div className={styles.BarChart}>
            <div className={styles.BarChart__title}>{props.title}</div>
            <YAxis
                title={'YAxis'}
                data={props.type === ChartType.StackedBarChart ? stackSeries(props.data) : props.data}
                config={props.yAxisConfig}
                padding={props.padding}
            />
            { renderPlot() }
            <XAxis
                title={'XAxis'}
                data={props.data}
                config={props.xAxisConfig}
                padding={props.padding}
            />
        </div>
    )
}

export default Chart;
