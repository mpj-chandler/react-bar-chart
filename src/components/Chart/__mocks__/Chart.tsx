import React from 'react';

const Chart: React.FC<any> = (props: any) => <div {...props}>Mock Chart {props.children}</div>

export default Chart;