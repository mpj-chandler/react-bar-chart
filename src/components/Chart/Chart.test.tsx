import React from 'react';
import TestRenderer from 'react-test-renderer';
import Chart, { BarChartProps } from './Chart';
import DataType from '../../enums/DataType';
import AnimationEasingType from '../../enums/AnimationEasingFunction';
import { DataPoint } from '../../__types__/seriesTypes';

jest.mock('../XAxis/XAxis');
jest.mock('../YAxis/YAxis');
jest.mock('../BarPlot/BarPlot.tsx');

describe('Bar Chart', () => {
    const props: BarChartProps = {
        title: 'anything',
        data: [
            {
                seriesName: 'A',
                type: {
                    x: DataType.NonNullNumeric,
                    y: DataType.NonNullNumeric,
                },
                points: [
                    {
                        x: 0,
                        y1: 10
                    },
                    {
                        x: 1,
                        y1: 20
                    },
                    {
                        x: 2,
                        y1: 30
                    }
                ]
            }
        ],
        padding: {
            top: 10,
            bottom: 10,
            left: 10,
            right: 10
        },
        fillFormatter: (seriesIndex: number, dataPoint: DataPoint, index: number) => { return '#000 '},
        animationEasingType: AnimationEasingType.None
    }
    it('it renders consistently', () => {
        const component = TestRenderer.create(<Chart {...props}/>)

        expect(component.toJSON()).toMatchSnapshot();
    })
})
