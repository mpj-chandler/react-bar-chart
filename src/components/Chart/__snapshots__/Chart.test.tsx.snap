// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar Chart it renders consistently 1`] = `
<div
  className="BarChart"
>
  <div
    className="BarChart__title"
  >
    anything
  </div>
  <div
    data={
      Array [
        Object {
          "points": Array [
            Object {
              "x": 0,
              "y1": 10,
            },
            Object {
              "x": 1,
              "y1": 20,
            },
            Object {
              "x": 2,
              "y1": 30,
            },
          ],
          "seriesName": "A",
          "type": Object {
            "x": "NON_NULL_NUMERIC",
            "y": "NON_NULL_NUMERIC",
          },
        },
      ]
    }
    padding={
      Object {
        "bottom": 10,
        "left": 10,
        "right": 10,
        "top": 10,
      }
    }
    title="YAxis"
  >
    Mock YAxis 
  </div>
  <svg
    className="BarPlot"
  >
    <g>
      <rect
        fill="#000 "
        height="40%"
        stroke="black"
        width="10%"
        x="5%"
        y="NaN%"
      />
      <rect
        fill="#000 "
        height="80%"
        stroke="black"
        width="10%"
        x="45%"
        y="NaN%"
      />
      <rect
        fill="#000 "
        height="120%"
        stroke="black"
        width="10%"
        x="85%"
        y="NaN%"
      />
    </g>
  </svg>
  <div
    data={
      Array [
        Object {
          "points": Array [
            Object {
              "x": 0,
              "y1": 10,
            },
            Object {
              "x": 1,
              "y1": 20,
            },
            Object {
              "x": 2,
              "y1": 30,
            },
          ],
          "seriesName": "A",
          "type": Object {
            "x": "NON_NULL_NUMERIC",
            "y": "NON_NULL_NUMERIC",
          },
        },
      ]
    }
    padding={
      Object {
        "bottom": 10,
        "left": 10,
        "right": 10,
        "top": 10,
      }
    }
    title="XAxis"
  >
    Mock XAxis 
  </div>
</div>
`;
